File: public/scoring.js
function normalizeText(text) {
  if (!text) return "";
  return String(text).toLowerCase().replace(/[^a-z0-9\s]/g, " ");
}

function finiteNumber(value) {
  if (value === null || value === undefined || value === "") return null;
  const number = typeof value === "number" ? value : Number(value);
  return Number.isFinite(number) ? number : null;
}

function degreeKey(value) {
  const text = String(value || "").toLowerCase();
  if (!text || text === "all") return "all";
  if (text.includes("doctor") || text.includes("phd")) return "phd";
  if (text.includes("master") || text.includes("msc") || text.includes("m.sc")) return "msc";
  if (text.includes("bachelor") || text.includes("bsc") || text.includes("b.sc")) return "bsc";
  return "unknown";
}

function hasAccessibleSource(sources) {
  const acceptedStatuses = new Set(["ok", "redirects", "pdf"]);
  const officialTypes = new Set([
    "official_program_page", "official_admission_page", "official_curriculum_page",
    "official_tuition_page", "official_scholarship_page", "official_department_page",
    "official_lab_page", "official_housing_page", "official_visa_or_government_page",
    "official_industry_partner_page"
  ]);
  return sources.some((source) => (
    officialTypes.has(String(source?.source_type || "").toLowerCase()) &&
    acceptedStatuses.has(String(source?.access_status || "").toLowerCase())
  ));
}

function hasVerifiedField(normalized, field) {
  return Array.isArray(normalized?.dataQuality?.verifiedFields) && normalized.dataQuality.verifiedFields.includes(field);
}

function hasVerifiedEnglishStudyOption(record, normalized) {
  if (!hasVerifiedField(normalized, "language")) return false;

  const languages = normalized?.teachingLanguage || [];
  const languageText = languages.join(" ").toLowerCase();
  if (/\benglish\b/.test(languageText)) return true;

  const languageProfile = record?.language_profile || {};
  const flags = record?.scoring_inputs?.hard_filter_flags || record?._scoring_inputs?.hard_filter_flags || {};
  if (languageProfile.english_study_option_available === true || flags.english_study_option_available === true) return true;

  // Legacy records used this flag for programmes such as TUM where an
  // English-completable specialisation exists even though the top-level
  // language value is "Mixed". A false value must not reject bilingual
  // records whose explicit language list contains English.
  if (languageText.includes("mixed") && flags.english_only_compatible === true) return true;
  return languageText.includes("mixed") && languageProfile.english_required === true;
}

function bounded(value, fallback = 50) {
  const number = finiteNumber(value);
  return number === null ? fallback : Math.min(100, Math.max(0, number));
}

const DEFAULT_SCORE_WEIGHTS = {
  academic_fit: 30,
  eligibility_language: 20,
  cost_funding: 20,
  career_research: 15,
  living_risk: 10,
  confidence_deadline: 5
};

function normalizeWeights(input = {}) {
  const sanitized = Object.fromEntries(Object.keys(DEFAULT_SCORE_WEIGHTS).map((key) => {
    const value = finiteNumber(input?.[key]);
    return [key, value === null ? 0 : Math.max(0, value)];
  }));
  const total = Object.values(sanitized).reduce((sum, value) => sum + value, 0);
  const basis = total > 0 ? sanitized : DEFAULT_SCORE_WEIGHTS;
  const basisTotal = Object.values(basis).reduce((sum, value) => sum + value, 0);
  return Object.fromEntries(Object.entries(basis).map(([key, value]) => [key, (value / basisTotal) * 100]));
}

function annualTuitionScore(value) {
  if (value === null) return 50;
  if (value <= 0) return 95;
  if (value <= 2000) return 85;
  if (value <= 5000) return 70;
  if (value <= 10000) return 55;
  if (value <= 20000) return 35;
  return 15;
}

function stringEntries(value) {
  if (Array.isArray(value)) {
    return value.map((item) => {
      if (typeof item === "string") return item.trim();
      if (!item || typeof item !== "object") return "";
      const candidate = item.en
        ?? item.name?.en
        ?? item.name
        ?? item.label?.en
        ?? item.label
        ?? item.title?.en
        ?? item.title;
      return typeof candidate === "string" ? candidate.trim() : "";
    }).filter(Boolean);
  }
  return typeof value === "string" && value.trim() ? [value] : [];
}

function uniqueDocumentedEntries(...values) {
  return [...new Set(values.flatMap(stringEntries).map((item) => normalizeText(item)).filter(Boolean))];
}

function documentedAcademicStrength(record, normalized, hasCurriculumEvidence, hasResearchEvidence) {
  const research = record?.research_profile || {};
  const curriculum = record?.curriculum_profile || {};
  const categories = normalized?.categoryProfile || {};
  const documentedTopics = new Set([
    ...stringEntries(categories.primary_categories),
    ...stringEntries(categories.secondary_categories),
    ...stringEntries(categories.subcategories),
    ...stringEntries(categories.normalized_tags),
    ...stringEntries(curriculum.specializations),
    ...stringEntries(research.research_focus_areas),
    ...stringEntries(research.department_research_areas)
  ].map((item) => normalizeText(item)).filter(Boolean));

  // A high result must be traceable to checked curriculum/research evidence.
  // Legacy prestige, ranking and self-entered numeric seeds are intentionally excluded.
  // The multipliers below are deliberately steep: two labs and eight labs are
  // very different research environments and must not land on the same score.
  let score = hasCurriculumEvidence ? 42 : 25;
  score += Math.min(12, documentedTopics.size * 1.5);
  if (!hasResearchEvidence) return bounded(score);

  const labs = uniqueDocumentedEntries(research.labs, research.named_facilities).length;
  const centers = uniqueDocumentedEntries(research.research_centers, research.key_institutes).length;
  const projects = stringEntries(research.space_or_aerospace_projects).length
    + stringEntries(research.satellite_or_flight_projects).length;
  const teams = stringEntries(research.student_teams).length;

  score += 10;
  score += Math.min(24, labs * 3);
  score += Math.min(6, centers * 2);
  score += Math.min(12, projects * 4);
  score += Math.min(4, teams * 1.5);
  return bounded(score);
}

function calculateScore(record, preferences, weights) {
  weights = normalizeWeights(weights);
  const explanation = [];
  const warnings = [];
  const profileMatch = {
    enabled: window.personalizationEnabled || false,
    personal_field_fit: null,
    matched_interests: [],
    profile_penalties: [],
    profile_boosts: []
  };
  const normalized = window.uniDataAdapter ? window.uniDataAdapter.normalizeUniversityRecord(record) : null;
  const profile = profileMatch.enabled ? window.userProfile : null;
  const categoryProfile = normalized?.categoryProfile || record.Category_Profile || {
    category_scores: {},
    subcategories: [],
    normalized_tags: []
  };
  const categoryScores = categoryProfile.category_scores || {};
  const subcategories = Array.isArray(categoryProfile.subcategories) ? categoryProfile.subcategories : [];
  const normalizedTags = Array.isArray(categoryProfile.normalized_tags) ? categoryProfile.normalized_tags : [];
  let passed = true;

  const requestedDegree = degreeKey(profile?.target_degree || preferences.degreeFilter || "All");
  const targetDegree = requestedDegree === "bsc" ? "msc" : requestedDegree;
  const recordDegree = degreeKey(normalized?.degreeLevel || normalized?.degree || record.Program_Degree);
  if (targetDegree !== "all" && recordDegree !== targetDegree) passed = false;

  const languages = normalized?.teachingLanguage || [];
  const languageText = languages.join(" ").toLowerCase();
  const hasEnglishStudyOption = hasVerifiedEnglishStudyOption(record, normalized);
  const hasOtherTeachingLanguage = /\b(german|french|dutch|italian|spanish|portuguese|swedish|japanese|korean|chinese|russian|turkish)\b/.test(languageText);
  const requiresEnglishOption = Boolean(preferences.onlyEnglish)
    || Boolean(profile && ["english_available", "english_only"].includes(profile.language_filter));
  if (requiresEnglishOption && !hasEnglishStudyOption) {
    passed = false;
    profileMatch.profile_penalties.push({ type: "language", reason: "No verified English-taught route or option is available." });
  }

  // Only a verified annual tuition value enters the affordability score and
  // hard cap. Semester fees, one-off enrolment fees and full-program totals
  // stay visible in details but are never mixed into the same numeric scale.
  const annualCost = hasVerifiedField(normalized, "tuition")
    ? finiteNumber(normalized?.tuitionPerYear)
    : null;
  const sliderCap = finiteNumber(preferences.maxTuition);
  const profileCap = finiteNumber(profile?.max_tuition_eur_per_year);
  const activeHardCap = sliderCap && sliderCap > 0
    ? sliderCap
    : (profile?.strict_budget && profileCap && profileCap > 0 ? profileCap : null);
  if (activeHardCap !== null && (annualCost === null || annualCost > activeHardCap)) {
    passed = false;
    profileMatch.profile_penalties.push({ type: "cost", reason: "Tuition is unknown or exceeds the selected hard limit." });
  }

  const hasCurriculumEvidence = hasVerifiedField(normalized, "curriculum");
  const hasResearchEvidence = hasVerifiedField(normalized, "research");
  const academicStrength = documentedAcademicStrength(record, normalized, hasCurriculumEvidence, hasResearchEvidence);
  let technicalMatch = hasCurriculumEvidence ? 60 : 40;
  if (profileMatch.enabled && Array.isArray(profile?.interests) && profile.interests.length > 0 && window.buildExpandedInterestProfile) {
    const expandedInterests = window.buildExpandedInterestProfile(profile.interests, window.INTEREST_GRAPH || {});
    let score = 0;
    let maxPossible = 0;
    const recordTags = new Set(normalizedTags);

    for (const [interestKey, userWeight] of expandedInterests.entries()) {
      maxPossible += userWeight;
      let matched = false;
      const legacyLabel = window.INTEREST_GRAPH?.[interestKey]?.label?.en || interestKey;
      if (recordTags.has(interestKey) || recordTags.has(legacyLabel) || subcategories.includes(interestKey) || subcategories.includes(legacyLabel)) {
        score += userWeight;
        matched = true;
      } else if (categoryScores[interestKey] !== undefined || categoryScores[legacyLabel] !== undefined) {
        const categoryScore = finiteNumber(categoryScores[interestKey] ?? categoryScores[legacyLabel]) || 0;
        score += userWeight * Math.max(0, Math.min(1, categoryScore / 100));
        matched = true;
      }
      if (matched) {
        profileMatch.matched_interests.push({
          interest_key: interestKey,
          match_strength: userWeight,
          match_type: profile.interests.some((interest) => interest.key === interestKey) ? "direct" : "graph_neighbor"
        });
      }
    }

    technicalMatch = maxPossible > 0 ? Math.round((score / maxPossible) * 100) : 40;
    if (!hasCurriculumEvidence) {
      technicalMatch = Math.min(technicalMatch, 40);
      warnings.push("Technical tags are not yet backed by a checked curriculum source.");
    }
    profileMatch.personal_field_fit = technicalMatch;
    explanation.push(technicalMatch >= 80
      ? "Strong personalized match with your interest graph."
      : technicalMatch >= 40
        ? "Partial match with your expanded interest fields."
        : "Weak match with your stated academic interests.");
  } else {
    const selectedCategories = preferences.selectedCategoryKeys || preferences.selectedKeywords || [];
    if (selectedCategories.length > 0) {
      let scoreSum = 0;
      for (const category of selectedCategories) {
        if (categoryScores[category] !== undefined) scoreSum += finiteNumber(categoryScores[category]) || 0;
        else if (subcategories.includes(category)) scoreSum += 100;
        else if (normalizedTags.includes(category)) scoreSum += 60;
      }
      technicalMatch = Math.min(100, Math.max(0, scoreSum / selectedCategories.length));
      if (!hasCurriculumEvidence) {
        technicalMatch = Math.min(technicalMatch, 40);
        warnings.push("Technical match is capped until the curriculum source is checked.");
      }
      explanation.push(technicalMatch >= 80
        ? "Strong academic match for your selected fields."
        : technicalMatch >= 40
          ? "Partial academic match for your selected fields."
          : "No strong match found for your selected fields.");
    } else {
      if (!hasCurriculumEvidence) warnings.push("Academic strength is conservative because curriculum evidence is incomplete.");
    }
  }
  // Programme-level relevance is an evidence-backed scope judgement. A well
  // documented aviation-management curriculum must not outrank direct space
  // engineering merely because it has many checked courses and research credits.
  const relevanceStatus = String(
    record?.program_profile?.relevance_status || record?.relevance_status || ""
  ).toLowerCase();
  const technicalRelevanceCaps = { strong: 100, medium: 70, weak: 25, needs_review: 100 };
  const academicRelevanceCaps = { strong: 100, medium: 78, weak: 45, needs_review: 100 };
  const technicalRelevanceCap = technicalRelevanceCaps[relevanceStatus] ?? 100;
  const academicRelevanceCap = academicRelevanceCaps[relevanceStatus] ?? 100;
  if (technicalMatch > technicalRelevanceCap) {
    technicalMatch = technicalRelevanceCap;
    warnings.push(`Technical fit is capped because programme relevance is classified as ${relevanceStatus}.`);
  }
  if (profileMatch.enabled) profileMatch.personal_field_fit = technicalMatch;
  // This slider is deliberately academic strength first. A selected technical
  // field can refine it, but cannot make an undocumented prestige claim win.
  const academicFit = Math.min(
    academicRelevanceCap,
    Math.round((academicStrength * 0.8) + (technicalMatch * 0.2))
  );
  explanation.push(`Academic strength ${academicStrength}/100 is based on checked curriculum and research evidence.`);
  if (preferences.minFieldFit && technicalMatch < preferences.minFieldFit) passed = false;

  let eligibilityFit = 55;
  if (hasEnglishStudyOption) {
    eligibilityFit += 30;
  } else if (hasOtherTeachingLanguage) {
    eligibilityFit -= 10;
  } else {
    warnings.push("No verified English study option is documented.");
  }
  const admissionMode = String(normalized?.admissionMode || "").toLowerCase();
  if (admissionMode.includes("direct")) {
    eligibilityFit += 20;
    explanation.push("Direct admission offers lower entry risk.");
  } else if (/(aptitude|committee|portfolio|interview|competitive)/.test(admissionMode)) {
    eligibilityFit -= profileMatch.enabled && profile?.admission_risk_tolerance === "low" ? 40 : 20;
    warnings.push(`Competitive admission process (${normalized.admissionMode}).`);
  }
  eligibilityFit = Math.min(100, Math.max(0, eligibilityFit));

  const tuitionScore = annualTuitionScore(annualCost);
  let scholarshipScore = 50;
  const scholarshipProfile = record.scholarship_profile || {};
  const scholarshipVerified = hasVerifiedField(normalized, "scholarship");
  if (scholarshipVerified && scholarshipProfile.non_eu_eligible === true) {
    scholarshipScore = 100;
    explanation.push("Scholarship eligibility is documented for non-EU students.");
  } else if (scholarshipVerified && scholarshipProfile.regional_scholarship_available === true) {
    scholarshipScore = 80;
    explanation.push("Regional scholarship availability is documented.");
  } else if (normalized?.scholarshipSummary) {
    warnings.push("Scholarship information needs verification.");
  }
  let costFit = (tuitionScore * 0.75) + (scholarshipScore * 0.25);
  if (annualCost === null) {
    warnings.push("Comparable verified annual tuition is unavailable, so tuition is neutral in the cost score.");
  } else if (annualCost > 10000) {
    warnings.push(`High yearly tuition (€${annualCost.toFixed(0)}).`);
  } else if (annualCost <= 2000) {
    explanation.push(`Low documented tuition (€${annualCost.toFixed(0)}/yr).`);
  }
  if (profileCap && annualCost !== null && annualCost > profileCap && !profile?.strict_budget) {
    costFit -= 30;
    if (profileMatch.enabled) profileMatch.profile_penalties.push({ type: "cost", reason: "Tuition exceeds the preferred maximum." });
  }
  costFit = Math.min(100, Math.max(0, costFit));

  let careerFit = hasResearchEvidence ? 50 : 40;
  const confirmedPartners = normalized?.confirmedPartners || [];
  const partnerText = normalizeText(confirmedPartners.map((partner) => typeof partner === "object" ? partner.name || partner.en || "" : partner).join(" "));
  const premiumPartners = ["esa", "dlr", "nasa", "jaxa", "airbus", "cern", "onera", "isae", "estec", "boeing", "safran", "thales", "leonardo", "ohb"];
  // Curated partner lists on a fully verified record are usable evidence even
  // when the dedicated "industry" flag was never set by the research pipeline.
  const partnersUsable = hasVerifiedField(normalized, "industry")
    || (normalized?.dataQuality?.status === "verified" && confirmedPartners.length > 0);
  if (partnersUsable) {
    const premiumMatches = premiumPartners.filter((partner) => partnerText.includes(partner)).length;
    careerFit += Math.min(12, confirmedPartners.length * 3);
    careerFit += Math.min(24, premiumMatches * 8);
    if (premiumMatches > 0) explanation.push("Documented partnerships with major aerospace organisations.");
  }
  if (hasResearchEvidence) {
    const careerProjects = stringEntries(record?.research_profile?.space_or_aerospace_projects).length
      + stringEntries(record?.research_profile?.satellite_or_flight_projects).length;
    careerFit += Math.min(10, careerProjects * 3);
  }
  if (normalized?.internshipMandatory === true && hasCurriculumEvidence) {
    careerFit += 8;
    explanation.push("A documented mandatory internship supports industry exposure.");
  }
  careerFit = Math.min(100, Math.max(0, careerFit));

  const housingVerified = hasVerifiedField(normalized, "housing");
  let livingFit = housingVerified ? 50 : 45;
  const cityCost = String(normalized?.cityCostLevel || normalized?.livingRisk || "").toLowerCase();
  if (housingVerified && cityCost.includes("very_high")) livingFit -= 30;
  else if (housingVerified && cityCost.includes("high")) livingFit -= 15;
  else if (housingVerified && cityCost.includes("low")) livingFit += 20;
  const housing = String(normalized?.housingDifficulty || "").toLowerCase();
  if (housingVerified && housing.includes("nightmare")) {
    livingFit -= profileMatch.enabled && profile?.housing_risk_tolerance === "low" ? 50 : 30;
    warnings.push("Housing market is extremely difficult.");
  } else if (housingVerified && housing.includes("very hard")) {
    livingFit -= profileMatch.enabled && profile?.housing_risk_tolerance === "low" ? 40 : 20;
    warnings.push("Housing market is very difficult.");
  } else if (housingVerified && housing.includes("hard")) {
    livingFit -= 10;
  } else if (housingVerified && (housing.includes("moderate") || housing.includes("easy"))) {
    livingFit += 20;
    explanation.push("Housing is relatively accessible.");
  }
  livingFit = Math.min(100, Math.max(0, livingFit));

  const sourceVerified = hasAccessibleSource(normalized?.sources || []);
  const qualityStatus = normalized?.dataQuality?.status || (normalized?.needsVerification ? "needs_verification" : "partial");
  const verifiedCount = normalized?.dataQuality?.verifiedFields?.length || 0;
  let confidenceFit = qualityStatus === "verified" ? 90 : qualityStatus === "partial" ? Math.min(75, 42 + (verifiedCount * 6)) : 20;
  if (!sourceVerified) confidenceFit = Math.min(confidenceFit, 20);
  if (qualityStatus !== "verified") {
    warnings.push("Critical data requires source verification.");
  } else {
    explanation.push("The record has checked official source evidence for all critical fields.");
  }

  const totalWeight = (
    weights.academic_fit + weights.eligibility_language + weights.cost_funding +
    weights.career_research + weights.living_risk + weights.confidence_deadline
  ) || 100;
  const finalScore = (
    academicFit * (weights.academic_fit / totalWeight) +
    eligibilityFit * (weights.eligibility_language / totalWeight) +
    costFit * (weights.cost_funding / totalWeight) +
    careerFit * (weights.career_research / totalWeight) +
    livingFit * (weights.living_risk / totalWeight) +
    confidenceFit * (weights.confidence_deadline / totalWeight)
  );

  const weightedComponents = {
    academic_fit: academicFit * (weights.academic_fit / totalWeight),
    eligibility_language: eligibilityFit * (weights.eligibility_language / totalWeight),
    cost_funding: costFit * (weights.cost_funding / totalWeight),
    career_research: careerFit * (weights.career_research / totalWeight),
    living_risk: livingFit * (weights.living_risk / totalWeight),
    confidence_deadline: confidenceFit * (weights.confidence_deadline / totalWeight)
  };

  return {
    passed_hard_filters: passed,
    total_score: finalScore,
    components: {
      academic_fit: academicFit,
      eligibility_language: eligibilityFit,
      cost_funding: costFit,
      career_research: careerFit,
      living_risk: livingFit,
      confidence_deadline: confidenceFit
    },
    weighted_components: weightedComponents,
    normalized_weights: weights,
    explanation,
    warnings,
    personalized_match: profileMatch
  };
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = { calculateScore, normalizeWeights };
} else {
  window.unirankScoring = { calculateScore, normalizeWeights };
}

File: unirank/core/scoring.py
# unirank/core/scoring.py
import re
from typing import Dict, Any, List, Tuple
from unirank.core.taxonomy import build_category_profile

def normalize_text(text: Any) -> str:
    if not text or str(text).lower() == 'nan':
        return ""
    # Remove non-alphanumeric and replace with space
    s = str(text).lower()
    s = re.sub(r'[^a-z0-9\s]', ' ', s)
    return s

def calculate_score(record: Dict[str, Any], preferences: Dict[str, Any], weights: Dict[str, float]) -> Dict[str, Any]:
    explanation = []
    warnings = []
    passed = True

    # Safely get values, handling pandas NaN or dict missing keys
    def get_val(keys):
        if isinstance(keys, str):
            keys = [keys]
        for k in keys:
            if k in record:
                v = record[k]
                if v is not None and str(v).lower() != 'nan':
                    return v
        return ""

    cat_profile = record.get('Category_Profile')
    if not cat_profile:
        cat_profile = build_category_profile(record)
        record['Category_Profile'] = cat_profile

    # --- HARD FILTERS ---
    
    # 1. Degree Filter
    target_degree = preferences.get('degreeFilter', 'All')
    rec_degree = str(get_val(['Program_Degree', 'program_degree'])).lower()
    if target_degree != 'All' and rec_degree:
        if target_degree.lower() not in rec_degree:
            passed = False

    # 2. English Only Filter
    lang_req = str(get_val(['Admission_Language_Req', 'admission_language_req'])).lower()
    if preferences.get('onlyEnglish', False):
        if 'german' in lang_req or 'french' in lang_req or 'dutch' in lang_req:
            passed = False

    # 3. Max Tuition
    tuit_raw = get_val(['tuition_eur_per_year', 'Tuition_Fee'])
    try:
        tuit = float(tuit_raw) if tuit_raw else 0.0
    except (ValueError, TypeError):
        tuit = 0.0
        
    max_tuit = preferences.get('maxTuition', 0)
    if max_tuit > 0:
        if tuit > max_tuit:
            passed = False

    # --- SCORE COMPONENTS ---
    
    # 1. Academic / Field Fit (0-100)
    academic_fit = 0.0
    selected_cats = preferences.get('selectedKeywords', [])
    
    if not selected_cats:
        academic_fit = 50.0
    else:
        score_sum = 0.0
        matches = 0
        for cat in selected_cats:
            # Is it a top category?
            if cat in cat_profile['category_scores']:
                cat_score = cat_profile['category_scores'][cat]
                score_sum += cat_score
                if cat_score >= 30:
                    matches += 1
            # Is it a subcategory?
            elif cat in cat_profile['subcategories']:
                score_sum += 100.0  # Perfect match for subcategory
                matches += 1
            elif cat in cat_profile['normalized_tags']:
                score_sum += 60.0 # Partial match
                matches += 1
        
        if len(selected_cats) > 0:
            academic_fit = score_sum / len(selected_cats)
            
        academic_fit = min(100.0, max(0.0, academic_fit))
        
        if academic_fit >= 80:
            explanation.append("Strong academic match for your selected fields.")
        elif academic_fit >= 40:
            explanation.append("Partial academic match for your selected fields.")
        else:
            explanation.append("No strong match found for your selected fields.")

    min_field_fit = preferences.get('minFieldFit', 0)
    if min_field_fit and academic_fit < min_field_fit:
        passed = False

    # 2. Eligibility & Language Fit (0-100)
    elig_fit = 70.0
    ad_mode = str(get_val(['Admission_Mode', 'admission_mode'])).lower()
    
    if 'english' in lang_req:
        elig_fit += 10.0
        if 'german' not in lang_req and 'french' not in lang_req and 'dutch' not in lang_req:
            elig_fit += 10.0
            explanation.append("Program is fully English-taught.")
        else:
            elig_fit -= 20.0
            warnings.append(f"Requires an additional language ({lang_req}).")
            
    if 'direct' in ad_mode:
        elig_fit += 20.0
        explanation.append("Direct admission offers lower entry risk.")
    elif any(x in ad_mode for x in ['aptitude test', 'committee', 'portfolio', 'interview']):
        elig_fit -= 20.0
        warnings.append(f"Competitive admission process ({ad_mode}).")
        
    elig_fit = min(100.0, max(0.0, elig_fit))

    # 3. Cost & Funding (0-100)
    max_t = 20000.0
    tuition_norm = min(1.0, tuit / max_t)
    tuition_score = (1.0 - tuition_norm) * 100.0
    
    sem_fee_raw = get_val(['semester_fee_eur', 'semester_fee'])
    try:
        sem_fee = float(sem_fee_raw) if sem_fee_raw else 0.0
    except (ValueError, TypeError):
        sem_fee = 0.0
        
    sem_fee_norm = min(1.0, sem_fee / 1000.0)
    sem_fee_score = (1.0 - sem_fee_norm) * 100.0
    
    scholarship_score = 0.0
    schol_info = get_val(['Scholarships_Info', 'scholarships_info'])
    if schol_info and len(schol_info) > 0:
        scholarship_score = 100.0
        explanation.append("Scholarships are available for non-EU students.")
        
    cost_fit = (tuition_score * 0.7) + (sem_fee_score * 0.2) + (scholarship_score * 0.1)
    cost_fit = min(100.0, max(0.0, cost_fit))
    
    if tuit > 10000:
        warnings.append(f"High yearly tuition (€{tuit:.0f}).")
    elif tuit <= 2000:
        explanation.append(f"Very affordable tuition (€{tuit:.0f}/yr).")

    # 4. Career / Research Ecosystem (0-100)
    career_fit = 50.0
    ecosystem_str = normalize_text(" ".join(str(x) for x in [
        get_val('Industry_Ecosystem'),
        get_val('Industry_Partners'),
        get_val('Analysis_Pros'),
        get_val('field_recognition')
    ]))
    
    premium_partners = ['esa', 'dlr', 'nasa', 'jaxa', 'airbus', 'cern', 'onera', 'isae', 'estec']
    partner_matches = sum(1 for p in premium_partners if p in ecosystem_str)
    
    career_fit += partner_matches * 15.0
    
    comp_int = str(get_val('Industry_Comp_Intensity')).lower()
    if 'high' in comp_int:
        career_fit += 10.0
        
    if get_val('Internship_Mandatory'):
        career_fit += 10.0
        explanation.append("Mandatory internship ensures industry exposure.")
        
    career_fit = min(100.0, max(0.0, career_fit))
    if partner_matches > 0:
        explanation.append("Strong aerospace/tech ecosystem (partners include major agencies/corps).")

    # 5. Living Risk (0-100)
    living_fit = 50.0
    cost_city = str(get_val('Cost_City_Living')).lower()
    
    if 'very_high' in cost_city:
        living_fit -= 30.0
    elif 'high' in cost_city:
        living_fit -= 15.0
    elif 'low' in cost_city:
        living_fit += 20.0
        
    housing = str(get_val('Living_Housing_Difficulty')).lower()
    if 'nightmare' in housing:
        living_fit -= 30.0
        warnings.append("Housing market is extremely difficult (Nightmare).")
    elif 'very hard' in housing:
        living_fit -= 20.0
        warnings.append("Housing market is very difficult.")
    elif 'hard' in housing:
        living_fit -= 10.0
    elif 'moderate' in housing or 'easy' in housing:
        living_fit += 20.0
        explanation.append("Housing is relatively accessible.")
        
    living_fit = min(100.0, max(0.0, living_fit))

    # 6. Data Confidence / Deadline (0-100)
    conf_fit = 80.0
    if get_val('Meta_Needs_Verification'):
        conf_fit -= 20.0
        warnings.append("Data needs manual verification.")
    else:
        explanation.append("Data is verified.")
        
    sources = get_val('Meta_Sources')
    if sources and len(sources) > 0:
        conf_fit += 20.0
        
    conf_fit = min(100.0, max(0.0, conf_fit))

    # --- WEIGHTED SUM ---
    w_total = sum([
        weights.get('academic_fit', 30),
        weights.get('eligibility_language', 20),
        weights.get('cost_funding', 20),
        weights.get('career_research', 15),
        weights.get('living_risk', 10),
        weights.get('confidence_deadline', 5)
    ]) or 100.0
    
    n_w = {
        'ac': weights.get('academic_fit', 30) / w_total,
        'el': weights.get('eligibility_language', 20) / w_total,
        'co': weights.get('cost_funding', 20) / w_total,
        'ca': weights.get('career_research', 15) / w_total,
        'li': weights.get('living_risk', 10) / w_total,
        'cf': weights.get('confidence_deadline', 5) / w_total
    }
    
    final_score = (
        (academic_fit * n_w['ac']) +
        (elig_fit * n_w['el']) +
        (cost_fit * n_w['co']) +
        (career_fit * n_w['ca']) +
        (living_fit * n_w['li']) +
        (conf_fit * n_w['cf'])
    )

    return {
        'passed_hard_filters': passed,
        'total_score': final_score,
        'components': {
            'academic_fit': academic_fit,
            'eligibility_language': elig_fit,
            'cost_funding': cost_fit,
            'career_research': career_fit,
            'living_risk': living_fit,
            'confidence_deadline': conf_fit
        },
        'explanation': explanation,
        'warnings': warnings
    }
